rust/datafusion/src/optimizer/boolean_comparison.rs

```diff
@@ -0,0 +1,270 @@
+                        right: Box::new(right),
+                    },
+                },
+                Operator::NotEq => match (&left, &right) {
+                    (Expr::Literal(ScalarValue::Boolean(b)), _) => match b {
+                        Some(false) | None => right,
+                        Some(true) => Expr::Not(Box::new(right)),
+                    },
+                    (_, Expr::Literal(ScalarValue::Boolean(b))) => match b {
+                        Some(false) | None => left,
+                        Some(true) => Expr::Not(Box::new(left)),
+                    },
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::NotEq,
+                        right: Box::new(right),
+                    },
+                },
+                _ => Expr::BinaryExpr {
+                    left: Box::new(left),
+                    op: op.clone(),
+                    right: Box::new(right),
+                },
+            }
+        }
+        Expr::Not(expr) => Expr::Not(Box::new(optimize_expr(&expr))),
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+        } => {
+            if expr.is_none() {
+                // recurse into CASE WHEN condition expressions
+                Expr::Case {
+                    expr: None,
+                    when_then_expr: when_then_expr
+                        .iter()
+                        .map(|(when, then)| (Box::new(optimize_expr(when)), then.clone()))
+                        .collect(),
+                    else_expr: else_expr.clone(),
+                }
+            } else {
+                // when base expression is specified, when_then_expr conditions are literal values
+                // so we can just skip this case
+                e.clone()
+            }
+        }
+        Expr::Alias { .. }
+        | Expr::Negative { .. }
+        | Expr::Column { .. }
+        | Expr::InList { .. }
+        | Expr::IsNotNull { .. }
+        | Expr::IsNull { .. }
+        | Expr::Cast { .. }
+        | Expr::ScalarVariable { .. }
+        | Expr::Between { .. }
+        | Expr::Literal { .. }
+        | Expr::ScalarFunction { .. }
+        | Expr::ScalarUDF { .. }
+        | Expr::AggregateFunction { .. }
+        | Expr::AggregateUDF { .. }
+        | Expr::Sort { .. }
+        | Expr::Wildcard => e.clone(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::logical_plan::{col, lit, LogicalPlanBuilder};
+    use crate::test::*;
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let mut rule = BooleanComparison::new();
+        let optimized_plan = rule.optimize(plan).expect("failed to optimize plan");
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+    }
+
+    #[test]
+    fn simplify_eq_expr() -> Result<()> {
```

Comment:
I think adding a test for rewriting expressions in a non-filter plan would be valuable (e.g. make a join plan or something)